supabase/migrations/20250604175753_send-swap-state-updates.sql

```diff
@@ -0,0 +1,261 @@
+
+    -- update the transaction state to completed
+    update wallet.transactions
+    set state = 'COMPLETED',
+        completed_at = now()
+    where id = v_swap.transaction_id;
+
+    return;
+end;
+$function$
+;
+
+
+-- replaces the old fail_cashu_token_swap function to handle case where token swap is failing a reversal transaction
+create or replace function wallet.fail_cashu_token_swap(
+    p_token_hash text, 
+    p_user_id uuid, 
+    p_swap_version integer, 
+    p_failure_reason text
+)
+returns void
+language plpgsql
+security invoker
+set search_path = ''
+as $function$
+declare
+    v_token_swap wallet.cashu_token_swaps;
+    v_reversed_transaction_id uuid;
+    v_send_swap wallet.cashu_send_swaps;
+begin
+    select * into v_token_swap
+    from wallet.cashu_token_swaps
+    where token_hash = p_token_hash and user_id = p_user_id;
+
+    if v_token_swap is null then
+        raise exception 'token swap for token hash % not found', p_token_hash;
+    end if;
+
+    if v_token_swap.state != 'PENDING' then
+        raise exception 'token swap for token hash % cannot be failed because it is not in pending state. current state: %', p_token_hash, v_token_swap.state;
+    end if;
+
+    -- special handling for "Token already claimed" failures
+    -- this handles the edge case where:
+    -- 1. user initiates a send swap reversal (creating a reversal transaction)
+    -- 2. receiver claims the token around the same time
+    -- 3. receiver's claim is processed first by the mint
+    -- 4. when the onSpent event triggers for the send swap, there's a related reversal transaction in pending state
+    -- 5. the reversal transaction will eventually fail with "Token already claimed"
+    -- 6. without this handling, the original send swap would stay in pending state forever
+    -- this ensures the original send swap is properly marked as completed when the reversal fails due to the token being claimed
+    if p_failure_reason = 'Token already claimed' then
+        -- get the reversed transaction id if this token swap is reversing a send transaction
+        select reversed_transaction_id into v_reversed_transaction_id
+        from wallet.transactions
+        where id = v_token_swap.transaction_id;
+
+        -- if this is reversing a send transaction, find and update the corresponding send swap
+        if v_reversed_transaction_id is not null then
+            select * into v_send_swap
+            from wallet.cashu_send_swaps
+            where transaction_id = v_reversed_transaction_id;
+
+            if v_send_swap is null then
+                raise exception 'no send swap found for reversed transaction %', v_reversed_transaction_id;
+            end if;
+
+            -- update send swap to completed with optimistic concurrency
+            update wallet.cashu_send_swaps
+            set state = 'COMPLETED',
+                version = version + 1
+            where id = v_send_swap.id and version = v_send_swap.version;
```

Comment:
we can just do for upadate here instead of version check. makes more sense.

every time where you are reading the record and using the read version in the same function, it makes more sense to use for update